### ISSUE TITLE
CPU problem with Restlet 2.0.15

### ISSUE THREAD

**Issue by gonzajg** [restlet/restlet-framework-java#1031]
Hi we are noticing a lot of activity in some threads related to Restlet. The problem starts when the servers are on production for a while. Then we take out all requests but still get a lot of CPU usage sporadically.

This is the top with threads option:

```
  PID USER      PR  NI  VIRT  RES  SHR S %CPU %MEM    TIME+  COMMAND
36774 apalabra  20   0 44.5g  11g  51m R 99.9 19.1 187:27.85 java
36780 apalabra  20   0 44.5g  11g  51m R 99.9 19.1 186:33.75 java
50351 apalabra  20   0 44.5g  11g  51m R 99.9 19.1  89:22.97 java
 2853 apalabra  20   0 44.5g  11g  51m R 99.9 19.1  82:24.84 java
```

When I do a jstack to the process and look for that specific threads I find these two stack traces:

```
Thread 50351: (state = IN_JAVA)
 - org.restlet.engine.io.ReaderInputStream.read(byte[], int, int) @bci=213, line=166 (Compiled frame; information may be imprecise)
 - java.io.InputStream.skip(long) @bci=44, line=222 (Compiled frame)
 - java.io.FilterInputStream.skip(long) @bci=5, line=151 (Interpreted frame)
 - org.restlet.engine.application.RangeInputStream.read(byte[], int, int) @bci=22, line=159 (Interpreted frame)
 - java.io.FilterInputStream.read(byte[]) @bci=5, line=107 (Compiled frame)
 - org.restlet.engine.io.BioUtils.copy(java.io.InputStream, java.io.OutputStream) @bci=8, line=75 (Compiled frame)
 - org.restlet.engine.application.RangeRepresentation.write(java.io.OutputStream) @bci=5, line=143 (Interpreted frame)
 - org.restlet.engine.http.ServerCall.writeResponseBody(org.restlet.representation.Representation, java.nio.channels.WritableByteChannel, java.io.OuputStream) @bci=18, line=553 (Compiled frame)
 - org.restlet.engine.http.ServerCall.sendResponse(org.restlet.Response) @bci=49, line=491 (Compiled frame [deoptimized])
 - org.restlet.ext.servlet.internal.ServletCall.sendResponse(org.restlet.Response) @bci=218, line=462 (Compiled frame [deoptimized])
 - org.restlet.engine.http.adapter.ServerAdapter.commit(org.restlet.engine.http.HttpResponse) @bci=442, line=200 (Compiled frame)
 - org.restlet.engine.http.HttpServerHelper.handle(org.restlet.engine.http.ServerCall) @bci=30, line=154 (Compiled frame)
 - org.restlet.ext.servlet.ServerServlet.service(javax.servlet.http.HttpServletRequest, javax.servlet.http.HttpServletResponse) @bci=24, line=1031 (ompiled frame)
 - javax.servlet.http.HttpServlet.service(javax.servlet.ServletRequest, javax.servlet.ServletResponse) @bci=366, line=790 (Compiled frame)
 - org.eclipse.jetty.servlet.ServletHolder.handle(org.eclipse.jetty.server.Request, javax.servlet.ServletRequest, javax.servlet.ServletResponse) @bc=89, line=808 (Compiled frame)
 - org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(javax.servlet.ServletRequest, javax.servlet.ServletResponse) @bci=211, line=1669 (Cmpiled frame)
 - org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(javax.servlet.http.HttpServletRequest, javax.servlet.http.HttpServletRespnse, javax.servlet.FilterChain) @bci=53, line=88 (Compiled frame)
 - org.springframework.web.filter.OncePerRequestFilter.doFilter(javax.servlet.ServletRequest, javax.servlet.ServletResponse, javax.servlet.FilterChan) @bci=383, line=76 (Compiled frame)
 - org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(javax.servlet.ServletRequest, javax.servlet.ServletResponse) @bci=103, line=1652 (Cmpiled frame)
 - org.eclipse.jetty.servlet.ServletHandler.doHandle(java.lang.String, org.eclipse.jetty.server.Request, javax.servlet.http.HttpServletRequest, java.servlet.http.HttpServletResponse) @bci=203, line=585 (Compiled frame)
 - org.eclipse.jetty.server.handler.ScopedHandler.handle(java.lang.String, org.eclipse.jetty.server.Request, javax.servlet.http.HttpServletRequest, avax.servlet.http.HttpServletResponse) @bci=32, line=143 (Compiled frame)
 - org.eclipse.jetty.security.SecurityHandler.handle(java.lang.String, org.eclipse.jetty.server.Request, javax.servlet.http.HttpServletRequest, java.servlet.http.HttpServletResponse) @bci=553, line=577 (Compiled frame)
```

```
Thread 2853: (state = IN_JAVA)
 - sun.nio.cs.Surrogate$Parser.parse(char, char[], int, int) @bci=104, line=242 (Compiled frame; information may be imprecise)
 - sun.nio.cs.UTF_8$Encoder.encodeArrayLoop(java.nio.CharBuffer, java.nio.ByteBuffer) @bci=264, line=542 (Compiled frame)
 - sun.nio.cs.UTF_8$Encoder.encodeLoop(java.nio.CharBuffer, java.nio.ByteBuffer) @bci=17, line=619 (Compiled frame)
 - java.nio.charset.CharsetEncoder.encode(java.nio.CharBuffer, java.nio.ByteBuffer, boolean) @bci=57, line=561 (Compiled frame)
 - org.restlet.engine.io.ReaderInputStream.read(byte[], int, int) @bci=116, line=146 (Compiled frame)
 - java.io.InputStream.skip(long) @bci=44, line=222 (Compiled frame)
 - java.io.FilterInputStream.skip(long) @bci=5, line=151 (Interpreted frame)
 - org.restlet.engine.application.RangeInputStream.read(byte[], int, int) @bci=22, line=159 (Interpreted frame)
 - java.io.FilterInputStream.read(byte[]) @bci=5, line=107 (Compiled frame)
 - org.restlet.engine.io.BioUtils.copy(java.io.InputStream, java.io.OutputStream) @bci=8, line=75 (Compiled frame)
 - org.restlet.engine.application.RangeRepresentation.write(java.io.OutputStream) @bci=5, line=143 (Interpreted frame)
 - org.restlet.engine.http.ServerCall.writeResponseBody(org.restlet.representation.Representation, java.nio.channels.WritableByteChannel, java.io.Ou
 - org.restlet.engine.http.ServerCall.sendResponse(org.restlet.Response) @bci=49, line=491 (Compiled frame [deoptimized])
 - org.restlet.ext.servlet.internal.ServletCall.sendResponse(org.restlet.Response) @bci=218, line=462 (Compiled frame [deoptimized])
 - org.restlet.engine.http.adapter.ServerAdapter.commit(org.restlet.engine.http.HttpResponse) @bci=442, line=200 (Compiled frame)
 - org.restlet.engine.http.HttpServerHelper.handle(org.restlet.engine.http.ServerCall) @bci=30, line=154 (Compiled frame)
 - org.restlet.ext.servlet.ServerServlet.service(javax.servlet.http.HttpServletRequest, javax.servlet.http.HttpServletResponse) @bci=24, line=1031 (
 - javax.servlet.http.HttpServlet.service(javax.servlet.ServletRequest, javax.servlet.ServletResponse) @bci=366, line=790 (Compiled frame)
```

Do you have any clues what could this be? I know we are using an outdated version of Restlet, but if we find that this issue is related to the project, we could update and fix it.

```
System specs
Jetty version: 9.2.8
Java version: 1.7.0_67, vendor: Oracle Corporation
Java home: /usr/java/jdk1.7.0_67/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "3.10.35-43.137.amzn1.x86_64", arch: "amd64", family: "unix"
```

Thanks!

**Comment by gonzajg** [comment]
Further information tells me that the code loops infinitely between these two traces:

"qtp1547352271-63616" prio=10 tid=0x00007f6544047000 nid=0xe7cf runnable [0x00007f6892ae8000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.cs.UTF_8$Encoder.encodeArrayLoop(UTF_8.java:542)
	at sun.nio.cs.UTF_8$Encoder.encodeLoop(UTF_8.java:619)
	at java.nio.charset.CharsetEncoder.encode(CharsetEncoder.java:561)
	at org.restlet.engine.io.ReaderInputStream.read(ReaderInputStream.java:146)
	at java.io.InputStream.skip(InputStream.java:222)
	at java.io.FilterInputStream.skip(FilterInputStream.java:151)
	at org.restlet.engine.application.RangeInputStream.read(RangeInputStream.java:159)
	at java.io.FilterInputStream.read(FilterInputStream.java:107)
	at org.restlet.engine.io.BioUtils.copy(BioUtils.java:75)

"qtp1547352271-63616" prio=10 tid=0x00007f6544047000 nid=0xe7cf runnable [0x00007f6892ae8000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.cs.UTF_8$Encoder.encodeArrayLoop(UTF_8.java:544)
	at sun.nio.cs.UTF_8$Encoder.encodeLoop(UTF_8.java:619)
	at java.nio.charset.CharsetEncoder.encode(CharsetEncoder.java:561)
	at org.restlet.engine.io.ReaderInputStream.read(ReaderInputStream.java:146)
	at java.io.InputStream.skip(InputStream.java:222)
	at java.io.FilterInputStream.skip(FilterInputStream.java:151)
	at org.restlet.engine.application.RangeInputStream.read(RangeInputStream.java:159)
	at java.io.FilterInputStream.read(FilterInputStream.java:107)
	at org.restlet.engine.io.BioUtils.copy(BioUtils.java:75)

**Comment by gonzajg** [comment]
Further information:

Both threads are in an infinite loop trying to resolve the same URI. The code loops in this while code of ReaderInputStream:

    @Override
    public int read(byte[] b, int off, int len) throws IOException {
        int result = 0;
        boolean iterate = true;

        while (iterate) {
            // Do we need to refill the byte buffer?
            if (!this.byteBuffer.hasRemaining() && !this.endReached) {
                // Do we need to refill the char buffer?
                if (!this.charBuffer.hasRemaining()) {
                    this.charBuffer.clear();
                    int read = this.reader.read(this.charBuffer);
                    this.charBuffer.flip();

                    if (read == -1) {
                        this.endReached = true;
                    }
                }

                if ((len > 0) && this.charBuffer.hasRemaining()) {
                    // Refill the byte buffer
                    this.byteBuffer.clear();
                    this.charsetEncoder.encode(this.charBuffer,
                            this.byteBuffer, this.endReached);
                    this.byteBuffer.flip();
                }
            }

            // Copies as much bytes as possible in the target array
            int readLength = Math.min(len, this.byteBuffer.remaining());

            if (readLength > 0) {
                this.byteBuffer.get(b, off, readLength);
                off += readLength;
                len -= readLength;
                result += readLength;
            }

            // Can we iterate again?
            iterate = (len > 0)
                    && (!this.endReached || this.byteBuffer.hasRemaining() || this.charBuffer
                            .hasRemaining());
        }

        if (this.endReached && (result == 0)) {
            result = -1;
        }

        return result;
    }

off = 1049
len = 999
len never changes, readLength is always 0

this.charsetEncoder.encode(this.charBuffer,this.byteBuffer, this.endReached) is giving CoderResult.UNDERFLOW

I'll keep investigating...

**Comment by gonzajg** [comment]
Going on and understanding the issue, I found that charBuffer is in position 1023 and capacity 1024 (and limit). When the method tries to encode the bytes in the charBuffer to "give" them to the byteBuffer, it returns UNDERFLOW and the position of the charBuffer is unchanged as so the byteBuffer. This results in a never ending read of the charBuffer.

Looking at the contents of the charBuffer I found that eclipse can't show the value of the element indexed 1022 as show in this image:

![image](https://cloud.githubusercontent.com/assets/1380026/6371213/28766bf2-bcd8-11e4-85cb-a6bbc134de4a.png)

In expression it shows no value either.

I don't understand though why does Jetty idle time plays no role on this and ends the request, and how can the bytes remaining in the charBuffer can't be read.

**Comment by santiagopoli** [comment]
+1, I'm having the same problem!

**Comment by gonzajg** [comment]
@thboileau any feedback on this?

**Comment by gonzajg** [comment]
Still fighting with the issue in our production servers. This conversation seems to denote a similar problem with charset decoder:
https://igniterealtime.jiveon.com/thread/33053

Specially: "the infinite loop can occur if the malformed sequence is exactly at the end of the buffer.  in this case, you need to skip, instead of backtrace."

**Comment by thboileau** [comment]
Thanks for the reference to the conversation. I have a look.

**Comment by thboileau** [comment]
Hello, I have a question for you, do you need the support of ranges in your application?

**Comment by gonzajg** [comment]
Is rangerepresentation used automatically by Jackson parser? We use Jackson's converter service and only serve JSON as result.

**Comment by gonzajg** [comment]
Here's a new similar problem using jackson binary representation, which we don't. It mentions handling underflow other way to avoid infinite loops.

https://github.com/michel-kraemer/bson4jackson/issues/35

**Comment by jlouvel** [comment]
Jackson requires the ConverterService, but not the RangeService which is to allow partial representations retrieval. Can you try to call getRangeService().setEnabled(false) in your Restlet Application constructor?

Also, if you could isolate the problem in a reproducible code sample/unit test that would make it much easier to fix for us.

**Comment by gonzajg** [comment]
Unfortunately we can't reproduce it. It only happens in production environments with high hit rate. 

I'm guessing the range representation is negotiated by the client?

I will try the code you suggested and get back. Thanks!

**Comment by gonzajg** [comment]
Hi Jerome,

The issue seems to be gone without range representation. Here's an image of 1 server (1st) with the fix and 2 without (2nd and 3rd). 

![image](https://cloud.githubusercontent.com/assets/1380026/6845943/7aa089f4-d398-11e4-8029-54ae6b526794.png)

First set of circles is when 2 threads get stuck in an infinite loop. Second set, you can see that even when the load is low, those 2 threads are still maxing 2 processors as opposed to the fixed server.

I wonder why did it start just some few weeks ago and never had it before. And also if I'm loosing something by disabling range representation. As far as we know we never did ask for range representation that we know of. Maybe android and iOS clients do it automatically.

I'm guessing the issue is related to the Jackson parser and range representation.

**Comment by thboileau** [comment]
If I understand well, the issue is to the fact that the ReaderInputStream fails to read some sequences of bytes and decode it properly. It came directly from the data that are read which may explain why it appears at some time.
Range representation is a quite useful feature, but it may have some side effects especially when converting bytes to characters since ranging applies to bytes, not to characters.
Having said that, I don't really understand why ReaderInputStream appears here.

**Comment by gonzajg** [comment]
Me neither. The issue is completely gone with range representations disabled. Do you want me to close the issue?

**Comment by thboileau** [comment]
@gonzajg , I would like to keep it open.

**Comment by amitchaturvedi** [comment]
What is the status of this Bug? Is it fixed?

**Comment by thboileau** [comment]
hi @amitchaturvedi,
we consider it has been fixed, but I keep it open to have a look at the presence of `ReaderInputStream`.

**Comment by bitsofinfo** [comment]
+1

**Comment by bitsofinfo** [comment]
are these all related?  we are seeing this on 2.2.2

is if fixed? in a version > 2.2.2?

http://restlet-discuss.1400322.n2.nabble.com/CPU-at-100-td3414274.html
https://github.com/restlet/restlet-framework-java/issues/794
